flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java

```diff
@@ -96,40 +97,51 @@ private int run(String[] args) throws FlinkException, CliArgsException {
 			kubernetesClusterClientFactory.createClusterDescriptor(configuration);
 
 		try {
-			final ClusterClient<String> clusterClient;
+			final ClusterClientProvider<String> clusterClientProvider;
 			String clusterId = kubernetesClusterClientFactory.getClusterId(configuration);
 			final boolean detached = !configuration.get(DeploymentOptions.ATTACHED);
-			final FlinkKubeClient kubeClient = KubeClientFactory.fromConfiguration(configuration);
+
+			boolean connectToExistingCluster = false;
+			try (final FlinkKubeClient kubeClient = KubeClientFactory.fromConfiguration(configuration)) {
+				connectToExistingCluster = clusterId != null && kubeClient.getRestService(clusterId).isPresent();
+			} catch (Exception e) {
+				LOG.info("Could not properly shutdown cluster client.", e);
+			}
 
 			// Retrieve or create a session cluster.
-			if (clusterId != null && kubeClient.getRestService(clusterId).isPresent()) {
-				clusterClient = kubernetesClusterDescriptor.retrieve(clusterId).getClusterClient();
```

Comment:
Why is calling `getClusterClient` a problem? Which call does the `RestClusterClient` do where it cannot reach the cluster?

Reply:
It's not the RestClusterClient that is the issue, but the retrieval of the REST API address and port (https://github.com/apache/flink/blob/master/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java#L96).

Reply:
Alright, this makes sense. Thanks for clarification.

Reply:
Hmm ok, I see the problem. Maybe the problem is that we need to create a `ClusterClient` in order to retrieve the `clusterId` even though this is something which is known before the `ClusterClient` is being created.

Reply:
I guess my point is that one could already retrieve the `clusterId` from the `ClusterClientProvider` given the current implementations because the `RestClusterClient` is being initialized with `clusterId` which it simply returns.

Reply:
hmm....is it a hard requirement that a cluster-id must be specified when starting a cluster, or can it also be auto-generated by Kubernetes?

If it is not required, then we could adjust the PR such that the cluster-id is only queried if `clusterId` is null: https://github.com/apache/flink/pull/13190/files#r472915726

Reply:
Actually we can make this change in any case.

Reply:
Turns the clusterId is never null. `KubernetesClusterClientFactory#createClusterDescriptor` ensures that the cluster-id is set in the configuration, so we actually never need the client here.

Reply:
It's a bit annoying to rely on it as is since it requires methods to be called in a specific order though,

Reply:
@tillrohrmann I've adjusted the code to explicitly set the cluster ID directly in the KubernetesSessionCli, and simplified the rest accordingly.